kong/tools/request_aware_table.lua

```diff
@@ -0,0 +1,122 @@
+--- NOTE: tool is designed to assist with **detecting** request contamination
+-- issues on CI, during test runs. It does not offer security safeguards.
+
+local table_new = require "table.new"
+local table_clear = require "table.clear"
+
+local debug_mode   = kong.configuration.log_level == "debug"
+local error        = error
+local rawset       = rawset
+local setmetatable = setmetatable
+local get_phase    = ngx.get_phase
+local var          = ngx.var
+local log          = ngx.log
+
+
+local NGX_VAR_PHASES = {
+  set           = true,
+  rewrite       = true,
+  access        = true,
+  content       = true,
+  header_filter = true,
+  body_filter   = true,
+  log           = true,
+  balancer      = true,
+}
+
+-- Check if access is allowed for table, based on the request ID
+local function enforce_sequential_access(table)
+  if not NGX_VAR_PHASES[get_phase()] then
+    -- not in a request context: allow access and reset allowed request ID
+    rawset(table, "__allowed_request_id", nil)
```

Comment:
We're reusing `"__allowed_request_id"` string a few times. Maybe we could save it as a variable at the top of the file to prevent from typo errors? Also ease the change of it in future if necessary.

Also I was wondering whether it would make sense to name the action in this line like:

```lua
local ALLOWED_REQUEST_KEY = "__allowed_request_id"

local function reset_table_access(table)
  rawset(table, ALLOWED_REQUEST_KEY, nil)
end

local function lock_table_access(table, request_id)
  rawset(table, ALLOWED_REQUEST_KEY, curr_request_id)
end

local function get_locked_id(table)
  return rawget(table, ALLOWED_REQUEST_KEY)
end
```

... although it's quite possible it's just too much 😄

Reply:
I agree about the variable. I'd rather avoid introducing three one-line functions. While they might offer a more understandable interface, I believe that the simplicity of these operations doesn't justify the added complexity.